hudi-common/src/main/java/org/apache/hudi/common/model/BaseAvroPayload.java

```diff
@@ -85,13 +85,20 @@ public boolean canProduceSentinel() {
    */
   protected static boolean isDeleteRecord(GenericRecord genericRecord) {
     final String isDeleteKey = HoodieRecord.HOODIE_IS_DELETED_FIELD;
+    final String isDeleteKeyMeta = HoodieRecord.OPERATION_METADATA_FIELD;
     // Modify to be compatible with new version Avro.
     // The new version Avro throws for GenericRecord.get if the field name
     // does not exist in the schema.
-    if (genericRecord.getSchema().getField(isDeleteKey) == null) {
-      return false;
+    if (genericRecord.getSchema().getField(isDeleteKey) != null) {
+      Object deleteMarker = genericRecord.get(isDeleteKey);
+      return (deleteMarker instanceof Boolean && (boolean) deleteMarker);
     }
-    Object deleteMarker = genericRecord.get(isDeleteKey);
-    return (deleteMarker instanceof Boolean && (boolean) deleteMarker);
+
+    if (genericRecord.getSchema().getField(isDeleteKeyMeta) != null) {
+      Object deleteMarker = genericRecord.get(isDeleteKeyMeta);
+      return (deleteMarker != null && "D".equals(deleteMarker.toString()));
```

Comment:
Nice catch, can you write a test case for Spark then?

Reply:
I'm afraid we can not ignore the whole payload bytes while the `_hoodie_operation` is `D`, such as in Flink: https://github.com/apache/hudi/blob/014168b948a5a1ca88ea3e8cca213dfd46cef3f4/hudi-flink-datasource/hudi-flink/src/main/java/org/apache/hudi/table/format/mor/MergeOnReadInputFormat.java#L820, there are use cases when the record is a DELETE, it is still needs to be emiit representing a retraction msg.

So, we suggest to fix the `_hoodie_operation` meta field handing just in the Spark snapshot relation, just like what we do specifically in the inputformat of Flink source.

Reply:
> I'm afraid we can not ignore the whole payload bytes while the `_hoodie_operation` is `D`, such as in Flink:
> 
> https://github.com/apache/hudi/blob/014168b948a5a1ca88ea3e8cca213dfd46cef3f4/hudi-flink-datasource/hudi-flink/src/main/java/org/apache/hudi/table/format/mor/MergeOnReadInputFormat.java#L820
> 
> , there are use cases when the record is a DELETE, it is still needs to be emiit representing a retraction msg.
> So, we suggest to fix the `_hoodie_operation` meta field handing just in the Spark snapshot relation, just like what we do specifically in the inputformat of Flink source.

I don't think this affects the use of this method. And it is better to put this logic in the getInsertValue method of BaseAvroPayload

Reply:
It affects the usage because the `getInsertValue` returns nothing after your change, we still need the whole payload for retraction.